contracts/shapella-upgrade/ShapellaUpgradeTemplate.sol

```diff
@@ -0,0 +1,374 @@
+        router.grantRole(router.STAKING_MODULE_PAUSE_ROLE(), _locator.depositSecurityModule());
+        router.grantRole(router.STAKING_MODULE_RESUME_ROLE(), _locator.depositSecurityModule());
+        router.grantRole(router.REPORT_EXITED_VALIDATORS_ROLE(), _locator.accountingOracle());
+        router.grantRole(router.REPORT_REWARDS_MINTED_ROLE(), _locator.lido());
+
+        // TODO: grant WQ roles after the new WQ contract version arrived
+        // TODO: grant EB roles after the new EB contract version arrived
+    }
+
+    function _passAdminRoleFromTemplateToVoting() internal {
+        _transferOZAdminFromThisToVoting(_hashConsensusForValidatorExitBusOracle);
+        _transferOZAdminFromThisToVoting(_hashConsensusForAccountingOracle);
+        _transferOZAdminFromThisToVoting(_locator.burner());
+        _transferOZAdminFromThisToVoting(_locator.stakingRouter());
+        _transferOZAdminFromThisToVoting(_locator.accountingOracle());
+        _transferOZAdminFromThisToVoting(_locator.validatorsExitBusOracle());
+        _transferOZAdminFromThisToVoting(_locator.withdrawalQueue());
+
+        IOssifiableProxy(_locator.stakingRouter()).proxy__changeAdmin(_voting);
+        IOssifiableProxy(_locator.accountingOracle()).proxy__changeAdmin(_voting);
+        IOssifiableProxy(_locator.validatorsExitBusOracle()).proxy__changeAdmin(_voting);
+        IOssifiableProxy(_locator.withdrawalQueue()).proxy__changeAdmin(_voting);
+    }
+
+    function _verifyUpgrade() internal view {
+        require(IVersioned(_locator.legacyOracle()).getContractVersion() == 4, "INVALID_LO_VERSION");
+        require(IVersioned(_locator.accountingOracle()).getContractVersion() == 1, "INVALID_AO_VERSION");
+        require(IVersioned(_locator.stakingRouter()).getContractVersion() == 1, "INVALID_SR_VERSION");
+        require(IVersioned(_locator.validatorsExitBusOracle()).getContractVersion() == 1, "INVALID_EB_VERSION");
+        require(IVersioned(_locator.withdrawalQueue()).getContractVersion() == 1, "INVALID_WQ_VERSION");
+
+        require(IOssifiableProxy(address(_locator)).proxy__getAdmin() == _voting, "INVALID_LOCATOR_ADMIN" );
+        require(IOssifiableProxy(_locator.accountingOracle()).proxy__getAdmin() == _voting, "INVALID_AO_PROXY_ADMIN");
+        require(IOssifiableProxy(_locator.stakingRouter()).proxy__getAdmin() == _voting, "INVALID_SR_PROXY_ADMIN");
+        require(IOssifiableProxy(_locator.validatorsExitBusOracle()).proxy__getAdmin() == _voting, "INVALID_EB_PROXY_ADMIN");
+        require(IOssifiableProxy(_locator.withdrawalQueue()).proxy__getAdmin() == _voting, "INVALID_WQ_PROXY_ADMIN");
+
+        require(IPausableUntil(_locator.validatorsExitBusOracle()).isPaused(), "EB_NOT_PAUSED");
+        require(IPausableUntil(_locator.withdrawalQueue()).isPaused(), "WQ_NOT_PAUSED");
+        require(IPausableUntil(_locator.withdrawalQueue()).getResumeSinceTimestamp() == _hardforkTimestamp, "INCORRECT_WQ_RESUME_SINCE_TIMESTAMP");
```

Comment:
maybe the same for VEBOracle, but not sure about the exact timestamp

Reply:
added todo